crates/control/src/services/builder/mod.rs

```diff
@@ -0,0 +1,310 @@
+    #[error("failed to create source catalog file")]
+    CreateSource(#[source] std::io::Error),
+    #[error("processing task {task:?} failed")]
+    Task {
+        task: String,
+        #[source]
+        err: std::io::Error,
+    },
+    #[error("failed to persist build {id}")]
+    PersistBuild {
+        id: Id<Build>,
+        #[source]
+        src: crate::services::builds_root::BuildsRootError,
+    },
+    #[error("sqlite database error")]
+    Sqlite(#[from] rusqlite::Error),
+    #[error("control plane database error")]
+    SqlxError(#[from] sqlx::Error),
+}
+
+pub async fn serve_builds<F>(ctx: AppContext, shutdown: F) -> Result<(), Error>
+where
+    F: std::future::Future<Output = ()>,
+{
+    let shutdown = shutdown.fuse();
+    pin_mut!(shutdown);
+
+    let mut backoff = std::time::Duration::ZERO;
+
+    loop {
+        let sleep = tokio::time::sleep(backoff).fuse();
+        pin_mut!(sleep);
+
+        select! {
+            _ = shutdown => {
+                return Ok(())
+            }
+            _ = &mut sleep => (),
+        };
+
+        // Begin a |txn| which will scope a held advisory lock on `Build.id`.
+        let mut txn = ctx.db().begin().await?;
+
+        if let Some(build) = dequeue_build(&mut txn).await? {
+            let Build {
+                id,
+                account_id,
+                catalog,
+                created_at,
+                state: _,
+                updated_at,
+            } = build;
+
+            let tmpdir = tempfile::TempDir::new().map_err(|e| Error::CreateDir(e))?;
+            info!(%id, %account_id, tmpdir=?tmpdir.path(), %created_at, %updated_at, "processing dequeued build");
+
+            let (state, db_path) = process_build(id, catalog.unwrap().0, tmpdir.path()).await?;
+            info!(?state, "processed build");
+
+            ctx.put_builds()
+                .put_build(id, &db_path)
+                .await
+                .map_err(|src| Error::PersistBuild { id, src })?;
+            info!(%id, "put build");
+
+            update_build_state(&mut txn, id, state).await?;
+            txn.commit().await?;
+
+            backoff = std::time::Duration::ZERO;
+        } else {
+            debug!("serve_builds found no build to dequeue. Sleeping...");
+            backoff = std::time::Duration::from_secs(5);
+        }
+    }
+}
+
+async fn process_build(
+    id: Id<Build>,
+    catalog: models::Catalog,
+    tmp_dir: &Path,
+) -> Result<(State, PathBuf), Error> {
+    // We perform the build under a ./builds/ subdirectory, which is a
+    // specific sub-path expected by temp-data-plane underneath its
+    // working temporary directory. This lets temp-data-plane use the
+    // build database in-place.
+    let builds_dir = tmp_dir.join("builds");
+    std::fs::create_dir(&builds_dir).map_err(|err| Error::CreateDir(err))?;
+
+    // Write our catalog source file within the build directory.
+    std::fs::File::create(&builds_dir.join(&format!("{}.flow.yaml", id)))
```

Comment:
> I don't think it's safe to use std::fs operations in a Tokio task.

It's not a safety or correctness issue, but it can block the executor.

I dug into this a bit. The tokio adapters are [wrappers of the std library](https://docs.rs/tokio/latest/src/tokio/fs/create_dir.rs.html#49-52) which _claim_ (in docs) to merely hint to the tokio runtime that this is a blocking call. Go's runtime does a very similar thing where it starts the blocking operation on one of its even threads (called P's) but also associates a timer, and if it takes to long then it marks the thread as backround and moves tasks of that P to a new system thread. I thought that was what tokio was doing to... but it's not.

Instead it's _always_ spawning a new background thread for the operation, which is [quite a bit slower](https://github.com/tokio-rs/tokio/issues/3664) than just using the standard library blocking versions.

To be honest, I'm not really clear why developers should every _choose_ to use tokio::fs given this. I think users would be better off managing their own background tasks for blocking jobs in almost all cases...

...

In any case this is probably moot because I agree that spawn_blocking is going to be more appropriate for builder daemons. I believe it's not an issue currently because the top-level main() does a synchronous join over the future, so we never send it through tokio::spawn.

Reply:
> It's not a safety or correctness issue, but it can block the executor.

Yep, sorry for the imprecise language here. That's what I was implying, not a memory safety issue.

> I dug into this a bit. The tokio adapters are [wrappers of the std library](https://docs.rs/tokio/latest/src/tokio/fs/create_dir.rs.html#49-52) which _claim_ (in docs) to merely hint to the tokio runtime that this is a blocking call. Go's runtime does a very similar thing where it starts the blocking operation on the current event thread (called P's) but associates a timer, and if it takes to long then it marks the thread as backround and moves tasks of that P to a new system thread. I thought that was what tokio was doing to... but it's not.
> 
> Instead it's _always_ spawning a new background thread for the operation, which is [quite a bit slower](https://github.com/tokio-rs/tokio/issues/3664) than just using the standard library blocking versions.

Interesting. That's not what I'd expected it to be doing under the hood.

> To be honest, I'm not really clear why developers should ever _choose_ to use tokio::fs given this. I would think users would be better off managing their own background tasks for blocking jobs in almost all cases...

I agree with you here. I mean, I guess it's better to avoid blocking the executor by spawning yet another thread, but if you care, you probably do want to manage this yourself.

> In any case this is probably moot because I agree that spawn_blocking is going to be more appropriate for builder daemons. I believe it's not an issue currently because the top-level main() does a synchronous join over the future, so we never send it through tokio::spawn.

Yeah, I think we may still want to use `spawn_blocking` and then pass that handle to `try_join`. [There is a warning in the docs](https://docs.rs/tokio/latest/tokio/macro.try_join.html#runtime-characteristics) about making sure you spawn if you want parallelism or if one of the joined futures blocks. I think this would currently manifest itself as the Builder blocking and preventing Axum from handling new requests.